### PR TITLE
fix: replace datamask with key from new key value api

### PR DESCRIPTION
**Pull request by pkdotson**
### SUMMARY
This pr fixes the longurl problem in dashboard when a user has a lot of filters and shares the url with other users. The long copied url had the potential to break loading the dashboard. This pr uses the new keyvalue api for dashboard to replace the datamask content with secure hashed key. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

before

https://user-images.githubusercontent.com/17326228/145321730-ad7048c7-cd34-4b8c-b3a6-1d74f595bf22.mov

after

https://user-images.githubusercontent.com/17326228/145322047-cd7cc9d4-7af3-472d-bf9a-3b40cc1cc38a.mov

### TESTING INSTRUCTIONS
Go to a dashboard and create filters. Ensure that the nativefilters param has a hashed key when applying new filters and saving the dashboard url.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/state.ts [hunk 1]
```diff
@@ -88,7 +88,6 @@ export const useFilterUpdates = (
 ) => {
   const filters = useFilters();
   const dataMaskApplied = useNativeFiltersDataMask();
-
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/state.ts:91]: Could we remove the unnecessary change?

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/utils.ts [hunk 2]
```diff
@@ -69,7 +69,6 @@ export const checkIsApplyDisabled = (
 ) => {
   const dataSelectedValues = Object.values(dataMaskSelected);
   const dataAppliedValues = Object.values(dataMaskApplied);
-
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/utils.ts:72]: Could we remove the unnecessary change?

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/keyValue.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    .then(r => r.json.message)
+    .catch(err => err);
+
+export const createFilterKey = (dashId: string, value: string) =>
+  SupersetClient.post({
+    endpoint: `api/v1/dashboard/${dashId}/filter_state`,
+    jsonPayload: { value },
+  })
+    .then(r => r.json.key)
+    .catch(err => console.log('err', err));
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/keyValue.tsx:35]:
```suggestion
    .catch(err => {
      logging.error(err);
      return null;
    });
```

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/keyValue.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ */
+import { SupersetClient } from '@superset-ui/core';
+
+export const updateFilterKey = (dashId: string, value: string, key: string) =>
+  SupersetClient.put({
+    endpoint: `api/v1/dashboard/${dashId}/filter_state/${key}/`,
+    jsonPayload: { value },
+  })
+    .then(r => r.json.message)
+    .catch(err => err);
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/keyValue.tsx:27]:
```suggestion
    .then(r => r.response.ok)
    .catch(err => {
      logging.error(err);
      return null;
    });
```

**Review by zhaoyongjie — COMMENTED** [review]
Hey Philip. I have added some comments.If you have any questions, feel free ping me.

**Comment by zhaoyongjie** [comment]
BTW:
I found a potential security issues when I review this PR. The user can create a cache from this `put` method. This can constitute a memory attack.

We have to fix this issue in separate PR. @michael-s-molina @pkdotson 

https://user-images.githubusercontent.com/2016594/145674947-84b4d336-28b2-4f48-a882-cb91a82c6ee5.mov

File: superset-frontend/src/constants.ts [hunk 5]
```diff
@@ -37,7 +37,7 @@ export const URL_PARAMS = {
   },
   nativeFilters: {
     name: 'native_filters',
-    type: 'rison',
+    type: 'rison | string',
   },
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset-frontend/src/constants.ts:41]: I agree with @zhaoyongjie that we shold decouple these two to keep the interface clean and simplify downstream logic (avoid having to check the type of the object). To keep the name simple, I think we could just call the new one `filterState`, and that would then contain the key for either native filter state or filter set state. Ping @amitmiran137 and @simcha90 any opinions here?

**Review by villebro — COMMENTED** [review]
A few quick comments. Great work and seems to work nicely, but it's missing one key feature that we discussed with @michael-s-molina that's needed: when a user enters a dashboard with the native filter key, it should initialize a new key-value pair based on the contents of the original one. This is to ensure a user can have multiple tabs open based on the same key; when opening the new tab, the key will change, which will ensure that the state of the original tab is not overwritten. This is mostly an edge case for regular users, but an important one for public dashboards, where circulating the key to others will lead to the original state being overwritten if another anonymous user changes the filters.

File: superset-frontend/src/constants.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (pkdotson) [reply superset-frontend/src/constants.ts:41]: I would agree to separate this out.

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx [hunk 6]
```diff
@@ -40,7 +40,8 @@ import {
 import Loading from 'src/components/Loading';
 import { getInitialDataMask } from 'src/dataMask/reducer';
 import { URL_PARAMS } from 'src/constants';
-import replaceUndefinedByNull from 'src/dashboard/util/replaceUndefinedByNull';
+import { getUrlParam } from 'src/utils/urlUtils';
+// import replaceUndefinedByNull from 'src/dashboard/util/replaceUndefinedByNull';
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx:44]: remove this

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx [hunk 7]
```diff
@@ -187,28 +193,45 @@ const FilterBar: React.FC<FiltersBarProps> = ({
...
+        dataMaskKey = nativeFiltersCacheKey;
+      } else {
+        let filterType;
+        const isOldRison = getUrlParam(URL_PARAMS.nativeFilters);
+        if (typeof isOldRison === 'object') {
+          filterType = rison.encode(isOldRison);
+        } else filterType = dataMask;
+
+        dataMaskKey = await createFilterKey(dashboardId, filterType);
+        console.log('dataMask', dataMaskKey);
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx:226]: remove console

File: superset-frontend/src/dashboard/util/getDashboardUrl.ts [hunk 8]
```diff
@@ -48,12 +44,9 @@ export default function getDashboardUrl({
...
   }
-
-  if (dataMask) {
-    newSearchParams.set(
-      URL_PARAMS.nativeFilters.name,
-      rison.encode(replaceUndefinedByNull(dataMask)),
-    );
+  const dataMaskKey = getUrlParam(URL_PARAMS.nativeFiltersByCacheKey);
+  if (dataMaskKey) {
+    newSearchParams.set(URL_PARAMS.nativeFilters.name, dataMaskKey as string);
```
Comment (villebro) [inline superset-frontend/src/dashboard/util/getDashboardUrl.ts:49]: I think we should try to avoid storing two types of state in a single key (see my comment on `URL_PARAMS` above). To make this backward compatible, I'd propose the following:
1. check if native filter key is provided. If so, fetch the native filter state from that and remove the native filter rison from the search params.
2. if no native filter key is provided, check if native filter rison parameter is provided. If so, parse that and initialize the filter state with it. Also, remove the native filter rison string from the URL when doing so to clean up the URL.
3. If neither 1) or 2) are true, initialize a new filter key with an empty state (I believe this is the current functionality).

**Review by villebro — COMMENTED** [review]
Second pass comments. I also noticed that loading a dashboard with a filter key causes a 404 to the `superset/favstar/Dashboard/undefined/count` endpoint (the `undefined` there should be the dashboard id):

image

This doesn't happen when loading the dashboard without the `native_filters` query string:
image

File: superset-frontend/cypress-base/cypress/integration/dashboard/key_value.test.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      expect(typeof initialFilterKey).eq('string');
+    });
+  });
+
+  it('should have different key when pages reloads', () => {
+    cy.location().then(loc => {
+      const newFilterKey = loc.search.split('=');
+      expect(newFilterKey).not.equal(initialFilterKey);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (pkdotson) [reply superset-frontend/cypress-base/cypress/integration/dashboard/key_value.test.ts:48]: This key is always is loaded even when no nativefilter is assigned. This test is testing exactly that, if you reload the page it test that when you apply a new nativefilter that a new key is created. I will add some comments.

File: superset-frontend/src/constants.ts [hunk 10]
```diff
@@ -39,6 +39,10 @@ export const URL_PARAMS = {
     name: 'native_filters',
     type: 'rison',
   },
+  nativeFiltersByCacheKey: {
+    name: 'native_filters',
```
[the remark this answers is not in the document]
Comment (pkdotson) [reply superset-frontend/src/constants.ts:43]: No the other can possibly return an rison string if poeple are using the old url which I check first before doing any requests.

[the remark this answers is not in the document]
Comment (pkdotson) [reply superset-frontend/src/constants.ts:43]: Actually yes that might make better sense. I just need to check and make sure backend won't complain about that change.

File: superset-frontend/src/dashboard/util/getDashboardUrl.ts (hunk 8 above)
Comment (pkdotson) [reply superset-frontend/src/dashboard/util/getDashboardUrl.ts:49]: These case are already done in the DashboardPage and Filterbar.
dashboardpage
Screen Shot 2021-12-16 at 4 40 58 PM
filterbar
Screen Shot 2021-12-16 at 4 41 39 PM

Comment (pkdotson) [reply superset-frontend/src/dashboard/util/getDashboardUrl.ts:49]: 1-3 are now done.

**Review by villebro — COMMENTED** [review]
Thanks for the changes; this works super nicely, just a few last minor nits!

File: superset-frontend/src/dashboard/components/menu/ShareMenuItems/index.tsx [hunk 11]
```diff
@@ -21,6 +21,12 @@ import { useUrlShortener } from 'src/common/hooks/useUrlShortener';
 import copyTextToClipboard from 'src/utils/copy';
 import { t } from '@superset-ui/core';
 import { Menu } from 'src/common/components';
+import { getUrlParam } from 'src/utils/urlUtils';
+import { URL_PARAMS } from 'src/constants';
+import {
+  createFilterKey,
+  getFilterValue,
+} from '../../nativeFilters/FilterBar/keyValue';
```
Comment (rusackas) [inline superset-frontend/src/dashboard/components/menu/ShareMenuItems/index.tsx:29]: maybe use `src/...` path?

File: superset-frontend/src/dashboard/containers/DashboardPage.tsx [hunk 12]
```diff
@@ -153,16 +154,47 @@ const DashboardPage: FC = () => {
   }, [readyToRender]);
 
   useEffect(() => {
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/containers/DashboardPage.tsx:156]: a little code style improvement. 
1. We don't have to catch exceptions here(catch exception logic in the API layer) 
2. We can make a utility function, e.g.:`isCachedFilterKey`, `isFilterObject`

```suggestion
  useEffect(() => {
    async function getDataMaskApplied(obj) {
      await getFilterValue(id, obj);
    }

    let dataMaskFromUrl;
    if (id > 0 && getUrlParam(URL_PARAMS.nativeFilters)) {
      dataMaskFromUrl = getUrlParam(URL_PARAMS.nativeFilters);
    }
    if (id > 0 && getUrlParam(URL_PARAMS.nativeFiltersKey)) {
      dataMaskFromUrl = getDataMaskApplied(getUrlParam(URL_PARAMS.nativeFiltersKey));
    }
    if (readyToRender) {
      if (!isDashboardHydrated.current) {
        isDashboardHydrated.current = true;
        if (isFeatureEnabled(FeatureFlag.DASHBOARD_NATIVE_FILTERS_SET)) {
          // only initialize filterset once
          dispatch(getFilterSets(id));
        }
      }
      dispatch(
        hydrateDashboard(
          dashboard,
          charts,
          filterboxMigrationState,
          dataMaskFromUrl,
        ),
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [readyToRender, filterboxMigrationState]);
```

File: superset-frontend/src/dashboard/containers/DashboardPage.tsx [hunk 13]
```diff
@@ -153,16 +154,47 @@ const DashboardPage: FC = () => {
...
+            dataMaskFromUrl,
+          ),
+        );
       }
-      dispatch(hydrateDashboard(dashboard, charts, filterboxMigrationState));
+      return null;
     }
+    if (dashboard?.id) getDataMaskApplied();
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [readyToRender, filterboxMigrationState]);
```
Comment (zhaoyongjie) [inline superset-frontend/src/dashboard/containers/DashboardPage.tsx:199]: a little code style improvement. 
1. We don't have to catch exceptions here(catch exception logic in the API layer) 
2. We can make a utility function, e.g.:`isCachedFilterKey`, `isFilterObject`
3. Dashboard id default value is `0`. If the dashboard has an id greater than 0, it will be a validated id.

```suggestion
  useEffect(() => {
    async function getDataMaskApplied(obj) {
      await getFilterValue(id, obj);
    }

    let dataMaskFromUrl;
    if (id > 0 && getUrlParam(URL_PARAMS.nativeFilters)) {
      dataMaskFromUrl = getUrlParam(URL_PARAMS.nativeFilters);
    }
    if (id > 0 && getUrlParam(URL_PARAMS.nativeFiltersKey)) {
      dataMaskFromUrl = getDataMaskApplied(getUrlParam(URL_PARAMS.nativeFiltersKey));
    }
    if (readyToRender) {
      if (!isDashboardHydrated.current) {
        isDashboardHydrated.current = true;
        if (isFeatureEnabled(FeatureFlag.DASHBOARD_NATIVE_FILTERS_SET)) {
          // only initialize filterset once
          dispatch(getFilterSets(id));
        }
      }
      dispatch(
        hydrateDashboard(
          dashboard,
          charts,
          filterboxMigrationState,
          dataMaskFromUrl,
        ),
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [readyToRender, filterboxMigrationState]);
```

Comment (pkdotson) [reply superset-frontend/src/dashboard/containers/DashboardPage.tsx:199]: This block still needs to be wrapped in async function other wise we end up with race conditions where it doesn't get the prev datamask and return empty object.

**Review by zhaoyongjie — APPROVED** [review]
frontend code LGTM, but we have to polish the key-value backend codes in separate PR. When I input a random filter key, the server-side response 500.

image

**Review by villebro — APPROVED** [review]
LGTM, thanks for the hard work here!

**Comment by michael-s-molina** [comment]
> frontend code LGTM, but we have to polish the key-value backend codes in separate PR. When I input a random filter key, the server-side response 500.

@zhaoyongjie This has been fixed in https://github.com/apache/superset/pull/17878

**Comment by michael-s-molina** [comment]
Thank you for the hard work @pkdotson!
